### PR TITLE
Issue 2

### PR DESCRIPTION
**Pull request by Hiw**
image
image

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_image
+      @image = Image.find(params[:id])
+    end
+
+    # Only allow a trusted parameter "white list" through.
+    def image_params
```
Comment (eric199wen) [inline app/controllers/images_controller.rb]: Since it is generated from scaffold, make sure you understand this.

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  # POST /images
+  def create
+    @image = Image.new(image_params)
+
+    if @image.save
+      redirect_to @image, notice: 'Image was successfully created.'
+    else
+      render :new
+    end
+  end
```
Comment (eric199wen) [inline app/controllers/images_controller.rb:29]: I cannot tell if it is generated by the command or you added it, put your own implementation of `create` in a separate commit would be better.

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    get new_image_url
+    assert_response :success
+  end
+
+  test "should create image" do
+    assert_difference('Image.count') do
+      post images_url, params: { image: { url: @image.url } }
+    end
+
+    assert_redirected_to image_url(Image.last)
```
Comment (eric199wen) [inline test/controllers/images_controller_test.rb:23]: Is it also generated by the command?

File: app/views/images/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @images.each do |image| %>
+      <tr>
+        <td><%= image.url %></td>
+        <td><%= link_to 'Show', image %></td>
+        <td><%= link_to 'Edit', edit_image_path(image) %></td>
+        <td><%= link_to 'Destroy', image, method: :delete, data: { confirm: 'Are you sure?' } %></td>
```
Comment (eric199wen) [inline app/views/images/index.html.erb:19]: Normally we don't add codes we don't need at the moment, that is why I recommend not using rails generator for Ropes.

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -46,13 +44,14 @@ def destroy
...
-    end
+  # Use callbacks to share common setup or constraints between actions.
+  def set_image
+    @image = Image.find(params[:id])
+  end
+
+  # Only allow a trusted parameter "white list" through.
+  def image_params
+    params.require(:image).permit(:url)
+  end
```
Comment (eric199wen) [inline app/controllers/images_controller.rb:56]: Change on this file seems to be only correcting rubocop, it should be in its own commit.

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -5,40 +5,40 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
   end
 
-  test "should update image" do
+  test 'should update image' do
     patch image_url(@image), params: { image: { url: @image.url } }
     assert_redirected_to image_url(@image)
   end
 
-  test "should destroy image" do
+  test 'should destroy image' do
```
Comment (eric199wen) [inline test/controllers/images_controller_test.rb:41]: This is also fixing rubocop.

File: app/models/image.rb [hunk 7]
```diff
@@ -1,2 +1,30 @@
...
+  rescue StandardError => e
+    case e
+    when SocketError
+      errors.add(:url, "- Could not find #{url}. Make sure it's correct!")
+    when RestClient::NotFound
+      errors.add(:url, '- Could not fetch the given url :( Got HTTP 404')
+    else
+      errors.add(:url, '- Oops, something went wrong!')
+    end
+  end
```
Comment (eric199wen) [inline app/models/image.rb:25]: Seems like the test for this model is missing, it is tricky to write test for it since we need to stub the `RestClient.get`.

File: app/controllers/images_controller.rb [hunk 8]
```diff
@@ -1,23 +1,21 @@
 class ImagesController < ApplicationController
-  before_action :set_image, only: [:show, :edit, :update, :destroy]
+  before_action :set_image, only: %i[show edit update destroy]
```
Comment (eric199wen) [inline app/controllers/images_controller.rb:2]: We have a convention to add co-author in the commit message. You can put `Co-authored-by: name <name@example.com>` in your commit body

File: app/controllers/images_controller.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  def new
+    @image = Image.new
+  end
+
+  # GET /images/1/edit
+  def edit
+  end
+
+  # POST /images
+  def create
```
Comment (eric199wen) [inline app/controllers/images_controller.rb:21]: Like we discussed in standup, let's try to add things one by one manually.

**Comment by Hiw** [comment]
[Issue 2: Save Image Link](https://trello.com/c/fL1OdYmP/363-issue-2-save-image-link)

File: app/models/image.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    # check header OK
+    if res.code != 200
+      errors.add(:url, "- Could not fetch the given url :( Got HTTP #{res.code}")
+    # check MIME content type
+    elsif !res.headers[:content_type].start_with?('image')
+      errors.add(:url, '- Hey! Wait a second, the URL is not an image')
+    end
+  end
+
+  def determine_error(err)
```
Comment (eric199wen) [inline app/models/image.rb:16]: Seems like `determine_error` and `check_response` can be private methods.

File: app/controllers/images_controller.rb (hunk 8 above)
Comment (eric199wen) [reply app/controllers/images_controller.rb:2]: We don't have edit/update/destroy actions in this PR, so having this line is unnecessary.

File: app/views/images/_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+<%= form_with(model: image, local: true) do |form| %>
```
Comment (eric199wen) [inline app/views/images/_form.html.erb:1]: I thought we would like to use `simple_form`? But if it works then it is fine.

File: app/controllers/images_controller.rb (hunk 1 above)
Comment (Hiw) [reply app/controllers/images_controller.rb]: understood!

File: app/controllers/images_controller.rb [hunk 12]
```diff
@@ -1,6 +1,11 @@
 class ImagesController < ApplicationController
   before_action :set_image, only: %i[show edit update destroy]
 
+  # GET /images/
+  def index
+    redirect_to new_image_path
+  end
+
```
Comment (eric199wen) [inline app/controllers/images_controller.rb:8]: This should be in the next issue haha

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (Hiw) [reply test/controllers/images_controller_test.rb:20]: yeahhh

File: app/models/image.rb (hunk 7 above)
Comment (Hiw) [reply app/models/image.rb:25]: Mocked this

File: app/controllers/images_controller.rb (hunk 8 above)
Comment (Hiw) [reply app/controllers/images_controller.rb:2]: added!

File: test/fixtures/images.yml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+image_url:
+  url: https://www.example.com/image1.jpg
+
+not_url:
+  url: test
+
+text_url:
+  url: google.com
```
Comment (eric199wen) [inline test/fixtures/images.yml:10]: We don't need this file.

File: app/controllers/images_controller.rb (hunk 9 above)
Comment (Hiw) [reply app/controllers/images_controller.rb:18]: removed

File: config/routes.rb [hunk 14]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :images
```
Comment (eric199wen) [inline config/routes.rb:2]: `resources :images, only: %w[create new show]`.

**Review by eric199wen — APPROVED** [review]
